subxt/src/backend/mod.rs

```diff
@@ -392,121 +434,212 @@ mod test {
+                                let _ = sender.send(i).await;
+                            }
+                        }
+                        Message::Single(item) => sender.send(item).await.unwrap(),
+                    };
+                });
             }
         }
 
         struct Data {
-            request: MockDataTable,
-            subscription: Subscription,
+            data_table: MockDataTable,
+            subscription_channel: Option<Subscription>,
+            subscription_handlers: HashMap<String, SubscriptionHandler>,
+            method_handlers: HashMap<String, MethodHandler>,
+        }
+
+        impl Data {
+            async fn call_meth<'a>(
+                &'a mut self,
+                meth: &str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, Box<serde_json::value::RawValue>> {
+                let method = self.method_handlers.get(meth).unwrap_or_else(|| {
+                    panic!("no method named {} registered. Params: {:?}", meth, params)
+                });
+
+                (*method)(&mut self.data_table, &mut self.subscription_channel, params)
+            }
+
+            async fn call_subscription<'a>(
+                &'a mut self,
+                sub: &str,
+                params: Option<Box<serde_json::value::RawValue>>,
+            ) -> super::rpc::RawRpcFuture<'a, super::rpc::RawRpcSubscription> {
+                let sub = self.subscription_handlers.get(sub).unwrap_or_else(|| {
+                    panic!(
+                        "no subscription named {} registered. Params: {:?}",
+                        sub, params
+                    )
+                });
+
+                (*sub)(&mut self.data_table, &mut self.subscription_channel, params)
+            }
+        }
+        pub struct MockRpcBuilder {
+            data: Data,
+        }
+
+        impl MockRpcBuilder {
+            pub fn new() -> Self {
+                Self {
+                    data: Data {
+                        data_table: MockDataTable::new(),
+                        subscription_channel: None,
+                        subscription_handlers: HashMap::new(),
+                        method_handlers: HashMap::new(),
+                    },
+                }
+            }
+
+            pub fn add_method(mut self, method_name: &str, meth: MethodHandler) -> Self {
+                self.data.method_handlers.insert(method_name.into(), meth);
+                self
+            }
+
+            pub fn add_subscription(
+                mut self,
+                subscription_name: &str,
+                subscription_handler: SubscriptionHandler,
+            ) -> Self {
+                self.data
+                    .subscription_handlers
+                    .insert(subscription_name.into(), subscription_handler);
+                self
+            }
+
+            pub fn add_mock_data_from_iter<
```

Comment:
since this is the only way to add `mock_data`, why not removing the `from_iter`? 

```suggestion
            pub fn add_mock_data<
```

Reply:
renamed